[bp/1.28] buffer: separate the BufferFragement release and drain tracker (#28770)

soulxu opened the pull request:
Commit Message: [bp/1.28] buffer: separate the BufferFragement release and drain tracker (#28770)
Additional Description:
This the backport of https://github.com/envoyproxy/envoy/pull/28770

There is a requirement for the user space io handle and iouring io handle to move the write buffer to its own buffer and invoke the drain trackers. But becomes an issue for the BufferFragment, the BufferFragement's releasor is invoked by the drain tracker also. After moving with invoke drain trackers, the memory in the BufferFragment will be released too, which leads to the Buffer reference some memory already released.
Risk Level: high
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes https://github.com/envoyproxy/envoy/issues/28760
Related to issue: https://github.com/envoyproxy/envoy/issues/28395

soulxu commented:
cc @phlax @ggreenway

soulxu commented:
/retest

phlax commented:
verify/examples is currently broken on the branch - working on a fix to underlying issues

phlax reviewed (commented):
@soulxu did this fix have a changelog message?

generally if something is worth backporting then its worth documenting the fix/change

if not, can we add on `main` and add the commit here

soulxu commented:
/retest

soulxu commented:
@phlax do you know what happend for the example tests?

phlax commented:
> do you know what happened for the example tests? 1.28 and 1.26 failed on example tests, but 1.27 and 1.25 passed.

vaguely yes - altho trying to fix laterally - we can land this without it passing that test

phlax commented:
@soulxu if you remove the extraneous commit ill land this - branch prs get rebased so we just want the 2 commits

/wait

soulxu commented:
> @soulxu if you remove the extraneous commit ill land this - branch prs get rebased so we just want the 2 commits
> 
> /wait

got it, thanks! I rebased and removed the extraneous commit
